virtio-queue/src/descriptor_utils.rs

```
@@ -0,0 +1,884 @@
+                buffers: other,
+                bytes_consumed: 0,
+            })
+        } else if rem == 0 {
+            Ok(DescriptorChainConsumer {
+                buffers: VecDeque::new(),
+                bytes_consumed: 0,
+            })
+        } else {
+            Err(Error::SplitOutOfBounds(offset))
+        }
+    }
+}
+
+/// Provides high-level interface over the sequence of memory regions
+/// defined by readable descriptors in the descriptor chain.
+///
+/// Note that virtio spec requires driver to place any device-writable
+/// descriptors after any device-readable descriptors (2.6.4.2 in Virtio Spec v1.1).
+/// Reader will skip iterating over descriptor chain when first writable
+/// descriptor is encountered.
+#[derive(Clone)]
+pub struct Reader<'a, B = ()> {
+    buffer: DescriptorChainConsumer<'a, B>,
+}
+
+impl<'a, B: Bitmap + BitmapSlice + 'static> Reader<'a, B> {
+    /// Construct a new Reader wrapper over `desc_chain`.
+    pub fn new<M>(
+        mem: &'a GuestMemoryMmap<B>,
+        desc_chain: DescriptorChain<M>,
+    ) -> Result<Reader<'a, B>>
+    where
+        M: Deref,
+        M::Target: GuestMemory + Sized,
+    {
+        let mut total_len: usize = 0;
+        let buffers = desc_chain
+            .readable()
+            .map(|desc| {
+                // Verify that summing the descriptor sizes does not overflow.
+                // This can happen if a driver tricks a device into reading more data than
+                // fits in a `usize`.
+                total_len = total_len
+                    .checked_add(desc.len() as usize)
+                    .ok_or(Error::DescriptorChainOverflow)?;
+
+                let region = mem
+                    .find_region(desc.addr())
+                    .ok_or(Error::FindMemoryRegion)?;
+                let offset = desc
+                    .addr()
+                    .checked_sub(region.start_addr().raw_value())
+                    .unwrap();
```

Is this safe or we should return an error?

---

I think it is safe because in the precedent line, we ensure that `desc.addr()` is in `region`, i.e., `desc.addr()` is >= `region.start_addr()`.